tests/lean/run/2846.lean

```diff
@@ -0,0 +1,74 @@
+/-!
+# `delabConstWithSignature` avoids using inaccessible names
+-/
+
+/-!
+Defined without named arguments, prints without named arguments.
+-/
+/-- info: String.append : String → String → String -/
+#guard_msgs in #check String.append
+
+/-!
+The List argument is not named, it is not printed as a named argument.
+-/
+/-- info: List.length.{u_1} {α : Type u_1} : List α → Nat -/
+#guard_msgs in #check List.length
+
+/-!
+All arguments are named, all are printed as named arguments.
+-/
+/-- info: Nat.pos_pow_of_pos {n : Nat} (m : Nat) (h : 0 < n) : 0 < n ^ m -/
+#guard_msgs in #check Nat.pos_pow_of_pos
+
+/-!
+The hypothesis is not a named argument, so it's not printed as a named argument.
+-/
+def Nat.pos_pow_of_pos' {n : Nat} (m : Nat) : 0 < n → 0 < n ^ m := Nat.pos_pow_of_pos m
+
+/-- info: Nat.pos_pow_of_pos' {n : Nat} (m : Nat) : 0 < n → 0 < n ^ m -/
+#guard_msgs in #check Nat.pos_pow_of_pos'
+
+/-!
+Repetition of a named argument, only the first is printed as a named argument.
+-/
+def foo (n n : Nat) : Fin (n + 1) := 0
+
+/-- info: foo (n : Nat) : (n : Nat) → Fin (n + 1) -/
+#guard_msgs in #check foo
+
+/-!
+Repetition of a named argument, only the first is printed as a named argument.
+This is checking that shadowing is handled correctly (that's not the responsibility of
+`delabConstWithSignature`, but it assumes that the main delaborator will handle this correctly).
+-/
+def foo' (n n : Nat) (a : Fin ((by clear n; exact n) + 1)) : Fin (n + 1) := 0
+
+/-- info: foo' (n : Nat) : (n_1 : Nat) → Fin (n + 1) → Fin (n_1 + 1) -/
+#guard_msgs in #check foo'
+
+/-!
+Named argument after inaccesible name, still stays after the colon.
+But, it does not print using named pi notation since this is not a dependent type.
+-/
+def foo'' : String → (needle : String) → String := fun _ yo => yo
+
+/-- info: foo'' : String → String → String -/
+#guard_msgs in #check foo''
```

Comment:
@nomeata With your first example, `needle` doesn't appear as a named argument because it's not a dependent type. With some more work, we could make it appear, but everything after the `:` doesn't have a reliable name, so I'm not sure if that's worth doing.

Reply:
Could you show those names that don’t have a macro scope (`.hasMacroScopes`)? (Just blindly guessing here, I don't know my way around these things yet).

Reply:
Everything after the colon is pretty printed with the usual delaborator, so this would take some coordination with that.

If there were a pretty printer option that could cause pi types to print dependently, then this delaborator could annotate the pi types along the spine of this expression with that option, the ones that don't have macro scopes, before handing it off to be delaborated.

Reply:
Thanks for the explanation, then this is out of scope for this PR